KomaMRICore/src/simulation/Functors.jl

```diff
@@ -43,35 +43,16 @@ x = x |> cpu
 cpu(x) = fmap(x -> adapt(KA.CPU(), x), x, exclude=_isleaf)
 
 #MotionModel structs
-adapt_storage(::KA.GPU, x::NoMotion) = x
 adapt_storage(::KA.GPU, x::SimpleMotion) = x
-function adapt_storage(backend::KA.GPU, xs::ArbitraryMotion)
-    fields = []
-    for field in fieldnames(ArbitraryMotion)
-        if field in (:ux, :uy, :uz)
-            push!(fields, adapt(backend, getfield(xs, field)))
-        else
-            push!(fields, getfield(xs, field))
-        end
-    end
-    return KomaMRICore.ArbitraryMotion(fields...)
-end
 
 #Precision
 paramtype(T::Type{<:Real}, m) = fmap(x -> adapt(T, x), m)
 adapt_storage(T::Type{<:Real}, xs::Real) = convert(T, xs)
 adapt_storage(T::Type{<:Real}, xs::AbstractArray{<:Real}) = convert.(T, xs)
 adapt_storage(T::Type{<:Real}, xs::AbstractArray{<:Complex}) = convert.(Complex{T}, xs)
 adapt_storage(T::Type{<:Real}, xs::AbstractArray{<:Bool}) = xs
-adapt_storage(T::Type{<:Real}, xs::SimpleMotion) = SimpleMotion(paramtype(T, xs.types))
 adapt_storage(T::Type{<:Real}, xs::NoMotion) = NoMotion{T}()
-function adapt_storage(T::Type{<:Real}, xs::ArbitraryMotion) 
-    fields = []
-    for field in fieldnames(ArbitraryMotion)
-        push!(fields, paramtype(T, getfield(xs, field)))
-    end
-    return ArbitraryMotion(fields...)
-end
+adapt_storage(T::Type{<:Real}, xs::SimpleMotion) = SimpleMotion(paramtype(T, xs.types))
```

Comment:
Solved in [e040a0f](https://github.com/JuliaHealth/KomaMRI.jl/pull/408/commits/e040a0f0786b5dfa4cf971843e15ce12bc7b045c). 

I was able to remove adapt functions for SimpleMotion and just use `@functor`. For achieving this, I changed the type of SimpleMotion.types from `Array{<:SimpleMotionType{T}}` to `Tuple{Vararg{<:SimpleMotionType{T}}}`, since CUDA (as well as, I think, the rest of the backends) adapts by default any type of variable that inherits from `AbstractArray`. To avoid a custom adapt, we need to use a type that is not adapted by CUDA when doing `gpu(phantom, backend)`, and a `Tuple` is the solution. 
I also added to KomaMRIBase the package TupleTools.jl in order to sort SimpleMotionType's in `initialize_motion` without allocations (since Base module has not a defined method for sorting tuples). I think that even if you have to use this package, this solution is worthwhile to avoid using adapts.

For `NoMotion`, the custom adapt is still necessary, but I think that's not a big problem

Reply:
Apart from that, I defined a constructor for SimpleMotion, so we can avoid wrapping the simplemotion types into an array or a tuple:

```julia
motion = SimpleMotion(PeriodicRotation(1.0, 0.5, 45.0, 0.0, 0.0), Translation(0.1, 0.1, 0.1, 0.0, 1.0))
```